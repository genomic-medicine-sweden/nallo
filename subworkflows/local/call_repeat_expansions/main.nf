include { TRGT                                   } from '../../../modules/local/trgt'
include { SAMTOOLS_INDEX as SAMTOOLS_INDEX_TRGT  } from '../../../modules/nf-core/samtools/index/main'
include { SAMTOOLS_SORT as SAMTOOLS_SORT_TRGT    } from '../../../modules/nf-core/samtools/sort/main'
include { BCFTOOLS_SORT as BCFTOOLS_SORT_TRGT    } from '../../../modules/nf-core/bcftools/sort/main'
include { BCFTOOLS_INDEX as BCFTOOLS_INDEX_MERGE } from '../../../modules/nf-core/bcftools/index/main'
include { BCFTOOLS_MERGE                         } from '../../../modules/nf-core/bcftools/merge/main'

workflow CALL_REPEAT_EXPANSIONS {

    take:
    ch_bam_bai  // channel: [mandatory] [ val(meta), path(bam), path(bai) ]
    ch_fasta    // channel: [mandatory] [ val(meta), path(fasta) ]
    ch_fai      // channel: [mandatory] [ val(meta), path(fai) ]
    ch_trgt_bed // channel: [mandatory] [ val(meta), path(bed) ]

    main:
    ch_repeat_calls_vcf = Channel.empty()
    ch_versions         = Channel.empty()

    ch_bam_bai
        .map { meta, bam, bai -> [meta, bam, bai, meta.sex] }
        .set { ch_trgt_input }

    // Run TGRT
    TRGT ( ch_trgt_input, ch_fasta, ch_trgt_bed.map { it[1] } )

    // Sort and index bam
    SAMTOOLS_SORT_TRGT ( TRGT.out.bam, [[],[]] )
    SAMTOOLS_INDEX_TRGT(SAMTOOLS_SORT_TRGT.out.bam)

    // Sort and index bcf
    BCFTOOLS_SORT_TRGT(TRGT.out.vcf)

    BCFTOOLS_SORT_TRGT.out.vcf
        .join( BCFTOOLS_SORT_TRGT.out.tbi )
        .map { meta, bcf, csi -> [ [ id : 'multisample' ], bcf, csi ] }
        .groupTuple()
        .set{ ch_bcftools_merge_in }

    BCFTOOLS_MERGE ( ch_bcftools_merge_in, ch_fasta, ch_fai, [] )

    BCFTOOLS_INDEX_MERGE ( BCFTOOLS_MERGE.out.merged_variants )

    ch_versions = ch_versions.mix(TRGT.out.versions)
    ch_versions = ch_versions.mix(SAMTOOLS_SORT_TRGT.out.versions)
    ch_versions = ch_versions.mix(SAMTOOLS_INDEX_TRGT.out.versions)
    ch_versions = ch_versions.mix(BCFTOOLS_SORT_TRGT.out.versions)
    ch_versions = ch_versions.mix(BCFTOOLS_MERGE.out.versions)
    ch_versions = ch_versions.mix(BCFTOOLS_INDEX_MERGE.out.versions)

    emit:
    vcf      = BCFTOOLS_SORT_TRGT.out.vcf  // channel: [ val(meta), path(vcf) ]
    versions = ch_versions                 // channel: [ versions.yml ]
}

