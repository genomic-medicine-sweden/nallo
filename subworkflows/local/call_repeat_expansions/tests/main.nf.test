nextflow_workflow {

    name "Test Workflow CALL_REPEAT_EXPANSIONS"
    script "../main.nf"
    config "./nextflow.config"
    workflow "CALL_REPEAT_EXPANSIONS"

    setup {
            run("GUNZIP") {
                script "../../../../modules/nf-core/gunzip/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test' ],
                        file(params.pipelines_testdata_base_path + 'reference/hg38.test.fa.gz', checkIfExists: true)
                    ]
                    """
                }

            }

            run("SAMTOOLS_FAIDX") {
                script "../../../../modules/nf-core/samtools/faidx/main.nf"
                process {
                    """
                    input[0] = GUNZIP.out.gunzip
                    input[1] = [[],[]]
                    """
                }

            }
    }

    test("[bam, bai], fasta, fai, bed - TRGT") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:false, family_id: 'family', sex: 1 ], // meta map
                    file(params.pipelines_testdata_base_path + 'testdata/HG002_PacBio_Revio.bam', checkIfExists: true),
                    file(params.pipelines_testdata_base_path + 'testdata/HG002_PacBio_Revio.bam.bai', checkIfExists: true)
                ])
                input[1] = GUNZIP.out.gunzip
                input[2] = SAMTOOLS_FAIDX.out.fai
                input[3] = Channel.of([
                    [ id: 'pathogenic_repeats' ],
                    file(params.pipelines_testdata_base_path + 'reference/pathogenic_repeats.hg38.bed')
                ])
                input[4] = 'trgt'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.sample_bai.get(0).get(1).endsWith(".bai") },
                { assert snapshot(
                    path(workflow.out.sample_vcf.get(0).get(1)).vcf.variantsMD5,
                    path(workflow.out.family_vcf.get(0).get(1)).vcf.variantsMD5,
                    bam(workflow.out.sample_bam.get(0).get(1), stringency: 'silent').getReadsMD5(),
                    workflow.out.versions,
                ).match() }
            )
        }

    }

    test("[bam, bai], fasta, fai, bed - STRdust") {

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:false, family_id: 'family', sex: 1 ], // meta map
                    file(params.pipelines_testdata_base_path + 'testdata/HG002_ONT.bam', checkIfExists: true),
                    file(params.pipelines_testdata_base_path + 'testdata/HG002_ONT.bam.bai', checkIfExists: true)
                ])
                input[1] = GUNZIP.out.gunzip
                input[2] = SAMTOOLS_FAIDX.out.fai
                input[3] = Channel.of([
                    [ id: 'pathogenic_repeats' ],
                    file(params.pipelines_testdata_base_path + 'reference/pathogenic_repeats.hg38.bed')
                ])
                input[4] = 'strdust'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    path(workflow.out.sample_vcf.get(0).get(1)).vcf.summary,
                    path(workflow.out.family_vcf.get(0).get(1)).vcf.summary,
                    workflow.out.versions,
                ).match() }
            )
        }
    }

    test("[bam, bai], fasta, fai, bed - TRGT - stub") {

        options "-stub"

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:false, family_id: 'family', sex: 1 ], // meta map
                    file(params.pipelines_testdata_base_path + 'testdata/HG002_PacBio_Revio.bam', checkIfExists: true),
                    file(params.pipelines_testdata_base_path + 'testdata/HG002_PacBio_Revio.bam.bai', checkIfExists: true)
                ])
                input[1] = GUNZIP.out.gunzip
                input[2] = SAMTOOLS_FAIDX.out.fai
                input[3] = Channel.of([
                    [ id: 'pathogenic_repeats' ],
                    file(params.pipelines_testdata_base_path + 'reference/pathogenic_repeats.hg38.bed')
                ])
                input[4] = 'trgt'
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }

    }

    test("[bam, bai], fasta, fai, bed - STRdust - stub") {

        options '-stub'

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:false, family_id: 'family', sex: 1 ], // meta map
                    file(params.pipelines_testdata_base_path + 'testdata/HG002_ONT.bam', checkIfExists: true),
                    file(params.pipelines_testdata_base_path + 'testdata/HG002_ONT.bam.bai', checkIfExists: true)
                ])
                input[1] = GUNZIP.out.gunzip
                input[2] = SAMTOOLS_FAIDX.out.fai
                input[3] = Channel.of([
                    [ id: 'pathogenic_repeats' ],
                    file(params.pipelines_testdata_base_path + 'reference/pathogenic_repeats.hg38.bed')
                ])
                input[4] = 'strdust'
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
