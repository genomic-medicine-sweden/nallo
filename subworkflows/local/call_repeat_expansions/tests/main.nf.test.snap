{
    "[bam, bai], fasta, fai, bed": {
        "content": [
            "502b7befd528ce70b8269a8e4b9281df",
            "1de84bd070a82b37e6fac25d19ae604e",
            "65999ab8f2bc7841de8172468bf23ab6",
            [
                "versions.yml:md5,52272b464e62cb9e5d41622ea76cd070",
                "versions.yml:md5,78d86f9ddc871705ca029444a498322f",
                "versions.yml:md5,799b136592e7434ff7eb9ddcc70e7e41",
                "versions.yml:md5,b9424dde80b33e84164cc956a14aa459",
                "versions.yml:md5,d7098c270381ae9c3c54fd4224f50206"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-11-04T16:23:30.11542862"
    },
    "[bam, bai], fasta, fai, bed - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "project": "project",
                            "sex": 1
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "project"
                        },
                        "project.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "project": "project",
                            "sex": 1
                        },
                        "test.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "project": "project",
                            "sex": 1
                        },
                        "test.bam.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    "versions.yml:md5,52272b464e62cb9e5d41622ea76cd070",
                    "versions.yml:md5,78d86f9ddc871705ca029444a498322f",
                    "versions.yml:md5,799b136592e7434ff7eb9ddcc70e7e41",
                    "versions.yml:md5,b9424dde80b33e84164cc956a14aa459",
                    "versions.yml:md5,d7098c270381ae9c3c54fd4224f50206"
                ],
                "project_vcf": [
                    [
                        {
                            "id": "project"
                        },
                        "project.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "sample_bai": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "project": "project",
                            "sex": 1
                        },
                        "test.bam.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "sample_bam": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "project": "project",
                            "sex": 1
                        },
                        "test.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "sample_vcf": [
                    [
                        {
                            "id": "test",
                            "single_end": false,
                            "project": "project",
                            "sex": 1
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,52272b464e62cb9e5d41622ea76cd070",
                    "versions.yml:md5,78d86f9ddc871705ca029444a498322f",
                    "versions.yml:md5,799b136592e7434ff7eb9ddcc70e7e41",
                    "versions.yml:md5,b9424dde80b33e84164cc956a14aa459",
                    "versions.yml:md5,d7098c270381ae9c3c54fd4224f50206"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-11-04T16:23:42.483750233"
    }
}
